# REDES NEURONALES EN DATOS DE SOBREVIVIENTES DEL TITANIC

In [ ]:
#----------------------------------------------------------
# Se importan las librerias
#----------------------------------------------------------

import numpy as np
import pandas as pd
import sklearn
import warnings

## 1. Leer el archivo titanic.csv. Tenga en cuenta que los elementos del archivo están separados por ‘;’

In [ ]:
#----------------------------------------------------------
# Url de los datos
#----------------------------------------------------------

url= ("https://raw.githubusercontent.com/MavelSterling/Informe_Inteligencia_Artificial/main/titanic.csv")

#----------------------------------------------------------
# Se cargan los datos
#----------------------------------------------------------
datos = pd.read_csv(url, sep=";")
datos


survived                                               Name  pclass  \
0           0                            Braund, Mr. Owen Harris       3   
1           1  Cumings, Mrs. John Bradley (Florence Briggs Th...       1   
2           1                             Heikkinen, Miss. Laina       3   
3           1       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1   
4           0                           Allen, Mr. William Henry       3   
..        ...                                                ...     ...   
886         0                              Montvila, Rev. Juozas       2   
887         1                       Graham, Miss. Margaret Edith       1   
888         0           Johnston, Miss. Catherine Helen "Carrie"       3   
889         1                              Behr, Mr. Karl Howell       1   
890         0                                Dooley, Mr. Patrick       3   

        sex   age  sibsp  parch     fare    who  embark_town alone  
0      male  22.0      1      0   7.2500    man  Southampton    no  
1    female  38.0      1      0  71.2833  woman    Cherbourg    no  
2    female  26.0      0      0   7.9250  woman  Southampton   yes  
3    female  35.0      1      0  53.1000  woman  Southampton    no  
4      male  35.0      0      0   8.0500    man  Southampton   yes  
..      ...   ...    ...    ...      ...    ...          ...   ...  
886    male  27.0      0      0  13.0000    man  Southampton   yes  
887  female  19.0      0      0  30.0000  woman  Southampton   yes  
888  female   NaN      1      2  23.4500  woman  Southampton    no  
889    male  26.0      0      0  30.0000    man    Cherbourg   yes  
890    male  32.0      0      0   7.7500    man   Queenstown   yes  

[891 rows x 11 columns]

In [ ]:
#----------------------------------------------------------
# Información de la base de datos
#----------------------------------------------------------

datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   Name         891 non-null    object 
 2   pclass       891 non-null    int64  
 3   sex          891 non-null    object 
 4   age          714 non-null    float64
 5   sibsp        891 non-null    int64  
 6   parch        891 non-null    int64  
 7   fare         891 non-null    float64
 8   who          891 non-null    object 
 9   embark_town  889 non-null    object 
 10  alone        891 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 76.7+ KB


## 2. Eliminar el atributo Name usando el comando datos.drop(["Name"], axis = 1, inplace = True)

In [ ]:
#----------------------------------------------------------
# Eliminar el atributo Name
#----------------------------------------------------------

datos.drop(["Name"], axis = 1, inplace = True)

#----------------------------------------------------------
# Observar los datos sin el atributo Name
#----------------------------------------------------------
datos

survived  pclass     sex   age  sibsp  parch     fare    who  \
0           0       3    male  22.0      1      0   7.2500    man   
1           1       1  female  38.0      1      0  71.2833  woman   
2           1       3  female  26.0      0      0   7.9250  woman   
3           1       1  female  35.0      1      0  53.1000  woman   
4           0       3    male  35.0      0      0   8.0500    man   
..        ...     ...     ...   ...    ...    ...      ...    ...   
886         0       2    male  27.0      0      0  13.0000    man   
887         1       1  female  19.0      0      0  30.0000  woman   
888         0       3  female   NaN      1      2  23.4500  woman   
889         1       1    male  26.0      0      0  30.0000    man   
890         0       3    male  32.0      0      0   7.7500    man   

     embark_town alone  
0    Southampton    no  
1      Cherbourg    no  
2    Southampton   yes  
3    Southampton    no  
4    Southampton   yes  
..           ...   ...  
886  Southampton   yes  
887  Southampton   yes  
888  Southampton    no  
889    Cherbourg   yes  
890   Queenstown   yes  

[891 rows x 10 columns]

In [ ]:
#----------------------------------------------------------
# Información de las variables de la base de datos
#----------------------------------------------------------
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   who          891 non-null    object 
 8   embark_town  889 non-null    object 
 9   alone        891 non-null    object 
dtypes: float64(2), int64(4), object(4)
memory usage: 69.7+ KB


In [ ]:
#----------------------------------------------------------
# Variables categóricas (tipo object)
#----------------------------------------------------------
datos.select_dtypes(include=['object']).describe()

sex  who  embark_town alone
count    891  891          889   891
unique     2    3            3     2
top     male  man  Southampton   yes
freq     577  537          644   537

In [ ]:
#----------------------------------------------------------
# Variables cuantitativas (tipo entero)
#----------------------------------------------------------
datos.select_dtypes(include=['int64']).describe()

survived      pclass       sibsp       parch
count  891.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642    0.523008    0.381594
std      0.486592    0.836071    1.102743    0.806057
min      0.000000    1.000000    0.000000    0.000000
25%      0.000000    2.000000    0.000000    0.000000
50%      0.000000    3.000000    0.000000    0.000000
75%      1.000000    3.000000    1.000000    0.000000
max      1.000000    3.000000    8.000000    6.000000

Se hace un resumen de las variables cuantitativas (tipo entero), para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo

In [ ]:
#----------------------------------------------------------
# Variables cuantitativas (tipo real)
#----------------------------------------------------------
datos.select_dtypes(include=['float64']).describe()

age        fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

Se hace un resumen de las variables cuantitativas (tipo real), para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo

## 3. Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas

In [ ]:
#----------------------------------------------------------
# Libreria sklearn
#----------------------------------------------------------
from sklearn.model_selection import train_test_split 

N=len(datos) # cantidad de datos
cTrain=int(N*0.8) # 80% para entrenar y 20% para probar
cTest=N-cTrain 

print(N,cTrain,cTest)# cantidad de: datos completos, datos entrenados, datos para probar

train_data,test_data= sklearn.model_selection.train_test_split(datos, train_size=cTrain, test_size=cTest)

891 712 179


891 es el número de observaciones en la base de datos

712 es el 80% de las observaciones de la base de datos que se van a utilizar para entrenar los datos

179 es el 20% de las observaciones de la base de datos que se van a utilizar para probar el modelo de predicción

In [ ]:
#----------------------------------------------------------
# Dimensiones de los datos de entrenamiento
#----------------------------------------------------------
train_data.shape

(712, 10)

Los datos de entrenamientos son 712 observaciones y 10 variables

In [ ]:
#----------------------------------------------------------
# Mostrar 5 filas de los datos de entrenamiento
#----------------------------------------------------------

train_data.head(5)

survived  pclass     sex   age  sibsp  parch     fare    who  \
45          0       3    male   NaN      0      0   8.0500    man   
285         0       3    male  33.0      0      0   8.6625    man   
84          1       2  female  17.0      0      0  10.5000  woman   
413         0       2    male   NaN      0      0   0.0000    man   
491         0       3    male  21.0      0      0   7.2500    man   

     embark_town alone  
45   Southampton   yes  
285    Cherbourg   yes  
84   Southampton   yes  
413  Southampton   yes  
491  Southampton   yes

De los 712 observaciones se muestran 5 datos con sus variables

In [ ]:
#----------------------------------------------------------
# Estadísticas descriptivas datos de entrenamiento
#----------------------------------------------------------
train_data.describe()

survived      pclass         age       sibsp       parch        fare
count  712.000000  712.000000  578.000000  712.000000  712.000000  712.000000
mean     0.383427    2.301966   29.795415    0.507022    0.373596   32.438634
std      0.486563    0.838656   14.277736    1.068646    0.789010   51.652555
min      0.000000    1.000000    0.670000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   21.000000    0.000000    0.000000    7.895800
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.427100
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.068750
max      1.000000    3.000000   80.000000    8.000000    5.000000  512.329200

Se hace un resumen de las variables cuantitativas de los datos de entrenamientos, para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo.

## 4. Utilizar una estrategia para normalizar los datos



### Pipeline para los atributos categóricos

In [ ]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.preprocessing import OneHotEncoder 
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# 5 variables categoricas
cat_attribs = ['pclass','sex','who','embark_town','alone']

cat_pipeline = Pipeline([
("imputer", SimpleImputer(strategy="most_frequent")), 
("cat_encoder", OneHotEncoder(sparse=False))
])



### Pipeline para los atributos numéricos

In [ ]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler 
from sklearn.preprocessing import MinMaxScaler

# Variables númericas

num_attribs = ['pclass','age', 'fare', 'sibsp', 'parch']
num_pipeline = Pipeline([("imputer", SimpleImputer(strategy="median")),
                         ("scaler", StandardScaler())])

### Pipeline completo

In [ ]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.compose import ColumnTransformer

full_pipeline = ColumnTransformer([ ("num", num_pipeline, num_attribs), 
                                   ("cat", cat_pipeline, cat_attribs),])



In [ ]:
#----------------------------------------------------------
# Se transforman los datos entrenados 
# Variables independientes
#----------------------------------------------------------

X_train = full_pipeline.fit_transform(train_data)
X_train.shape


(712, 18)

Los datos de entrenamientos son 712 observaciones y 18 variables

In [ ]:
X_train[0,:]

array([ 0.83290956, -0.11322937, -0.47249898, -0.47478674, -0.47383197,
        0.        ,  0.        ,  1.        ,  0.        ,  1.        ,
        0.        ,  1.        ,  0.        ,  0.        ,  0.        ,
        1.        ,  0.        ,  1.        ])

In [ ]:
#----------------------------------------------------------
# Variable dependiente -> survived
#----------------------------------------------------------

y_train = train_data["survived"]
y_train

45     0
285    0
84     1
413    0
491    0
      ..
243    0
456    0
183    1
368    1
511    0
Name: survived, Length: 712, dtype: int64

In [ ]:
X_test = full_pipeline.transform(test_data) 
X_test

array([[-1.55353553,  2.06199666, -0.11408485, ...,  1.        ,
         0.        ,  1.        ],
       [ 0.83290956, -0.26860266, -0.34842569, ...,  0.        ,
         1.        ,  0.        ],
       [ 0.83290956, -0.26860266, -0.47564721, ...,  1.        ,
         0.        ,  1.        ],
       ...,
       [-0.36031299,  0.58595042, -0.1247404 , ...,  1.        ,
         1.        ,  0.        ],
       [ 0.83290956, -0.57934924, -0.47548641, ...,  1.        ,
         0.        ,  1.        ],
       [ 0.83290956, -0.26860266,  0.46607719, ...,  1.        ,
         0.        ,  1.        ]])

In [ ]:
y_test = test_data["survived"]
y_test

467    0
73     0
810    0
633    0
382    0
      ..
195    1
1      1
594    0
521    0
509    1
Name: survived, Length: 179, dtype: int64

## 5. Construir 5 redes neuronales variando en la topología de la red la cantidad de capas ocultas y de neuronas por cada capa oculta. Puede también variar los hiperparámetros solver y la función de activación. En todas las pruebas debe usar un random_state=123. Incluya en el notebook una tabla a manera de resumen con el accuracy obtenido en cada caso y también las matrices de confusión

In [46]:
#----------------------------------------------------------
# Librerias
#----------------------------------------------------------
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Se crea una lista para almacenar los resultados
resultados = []

# Define una lista de topologías a probar
topologias = [
    (10,), # una capa oculta con 10 neuronas
    (10, 10), # dos capas ocultas con 10 neuronas cada una
    (20,), # una capa oculta con 20 neuronas
    (20, 20), # dos capas ocultas con 20 neuronas cada una
    (30,), # una capa oculta con 30 neuronas
    ]

# Define una lista de funciones de activación a probar
func_activacion = ["relu", "logistic"] 

# Define una lista de solvers a probar
solvers = ["adam", "sgd"]

# Itera para cada topología
for topologia in topologias:
    # Itera para cada función de activación
    for activacion in func_activacion:
        # Itera para cada solver
        for solver in solvers:
            # Crea el modelo de red neuronal
            model = MLPClassifier(hidden_layer_sizes=topologia, activation=activacion, solver=solver, random_state=123)
            # Entrena el modelo
            model.fit(X_train, y_train)
            # Realiza las predicciones
            y_pred = model.predict(X_test)
            # Calcula el accuracy
            accuracy = round(accuracy_score(y_test, y_pred),4) # con 4 decimales
            # Calcula la matriz de confusión
            confusion = confusion_matrix(y_test, y_pred)
            # Almacena el resultado
            resultados.append({"Topología": topologia, "Función de activación": activacion, "Solver": solver, "Accuracy": accuracy, "Matriz de confusión": confusion})

# Imprime los resultados
#print(resultados)


/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:

In [47]:
#----------------------------------------------------------
#Libreria pandas
#----------------------------------------------------------
import pandas as pd

# Se imprime la tabla con los resultados
result = pd.DataFrame(resultados)
display(result)

Topología Función de activación Solver  Accuracy    Matriz de confusión
0      (10,)                  relu   adam    0.8212   [[98, 12], [20, 49]]
1      (10,)                  relu    sgd    0.8045   [[95, 15], [20, 49]]
2      (10,)              logistic   adam    0.8156   [[96, 14], [19, 50]]
3      (10,)              logistic    sgd    0.6145    [[110, 0], [69, 0]]
4   (10, 10)                  relu   adam    0.8212   [[99, 11], [21, 48]]
5   (10, 10)                  relu    sgd    0.8324   [[99, 11], [19, 50]]
6   (10, 10)              logistic   adam    0.8156   [[97, 13], [20, 49]]
7   (10, 10)              logistic    sgd    0.6145    [[110, 0], [69, 0]]
8      (20,)                  relu   adam    0.8212  [[100, 10], [22, 47]]
9      (20,)                  relu    sgd    0.8156   [[98, 12], [21, 48]]
10     (20,)              logistic   adam    0.8156   [[96, 14], [19, 50]]
11     (20,)              logistic    sgd    0.6201    [[110, 0], [68, 1]]
12  (20, 20)                  relu   adam    0.8212   [[101, 9], [23, 46]]
13  (20, 20)                  relu    sgd    0.8045   [[96, 14], [21, 48]]
14  (20, 20)              logistic   adam    0.8212   [[98, 12], [20, 49]]
15  (20, 20)              logistic    sgd    0.6145    [[110, 0], [69, 0]]
16     (30,)                  relu   adam    0.8268   [[101, 9], [22, 47]]
17     (30,)                  relu    sgd    0.8156   [[96, 14], [19, 50]]
18     (30,)              logistic   adam    0.8156   [[96, 14], [19, 50]]
19     (30,)              logistic    sgd    0.6201    [[109, 1], [67, 2]]

Con los resultados obtenidos se construyó la tabla anterior, en la cual se tiene la topología de la red, la función de activación, el hiperparámetro solver, el accuracy y la matriz de confusión.

### 6. Indique en el notebook usando una celda de tipo Markdown los hiperparámetros que por el momento le permiten obtener la red con mayor accuracy

In [48]:
# Encontrar la red neuronal con el mayor valor de Accuracy
max_accuracy = max(resultados, key=lambda x: x["Accuracy"])

# Imprime los hiperparámetros del mejor resultado
print("-------------------------------------------")
print("Mejores hiperparámetros:")
print("-------------------------------------------")
print("Topología:", max_accuracy["Topología"])
print("Función de activación:", max_accuracy["Función de activación"])
print("Solver:", max_accuracy["Solver"])
print("Accuracy:", max_accuracy["Accuracy"])
print("Matriz de confusión:", max_accuracy["Matriz de confusión"])

-------------------------------------------
Mejores hiperparámetros:
-------------------------------------------
Topología: (10, 10)
Función de activación: relu
Solver: sgd
Accuracy: 0.8324
Matriz de confusión: [[99 11]
 [19 50]]


Se obtiene la red neuronal con el valor más alto en accuracy y se imprime los hiperparámetros.

### 7. Seleccione uno de los hiperparámetros disponibles en la documentación (https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html) que sea diferente al solver, a la función de activación, y al random_state. Realice dos variaciones en el hiperparámetro seleccionado manteniendo los otros hiperparámetros del punto anterior. Indique el accuracy obtenido al modificar el hiperparámetro seleccionado y analice si la red mejora, empeora, o mantiene su exactitud. Incluya en el notebook dicho análisis




In [93]:
#-----------------------------------------------
# Librerias
#-----------------------------------------------
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

# Se crea una lista para almacenar los resultados
resultados_p7 = []

# Define una lista de tasas de aprendizaje inicial para la prueba
learning_rate_inits = [0.01, 0.001]
activation = "relu" # función de activación ReLU
solver = "adam" # solver Adam
# Itera para cada tasa de aprendizaje inicial
for learning_rate_init in learning_rate_inits:

    # Define la topología de la red
    hidden_layer_sizes = (10, 10) # dos capas ocultas con 10 neuronas cada una
    # Crea el modelo de red neuronal
    model = MLPClassifier(hidden_layer_sizes=hidden_layer_sizes, activation=activation, solver=solver, random_state=123, learning_rate_init=learning_rate_init)
    # Entrena el modelo
    model.fit(X_train, y_train) 
    # Realiza las predicciones
    y_pred = model.predict(X_test)
    # Calcula el accuracy
    accuracy = accuracy_score(y_test, y_pred)
    # Almacena el resultado
    resultados_p7.append({"learning_rate_init": learning_rate_init, "accuracy": accuracy})

# Imprime los resultados
print("---------------------------------")
print(resultados_p7)
print("---------------------------------")

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


---------------------------------
[{'learning_rate_init': 0.01, 'accuracy': 0.8379888268156425}, {'learning_rate_init': 0.001, 'accuracy': 0.8212290502793296}]
---------------------------------


/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [94]:
#----------------------------------------------------------
#Libreria pandas
#----------------------------------------------------------
import pandas as pd

# Se imprime la tabla con los resultados
resultado_punto7 = pd.DataFrame(resultados_p7)
display(resultado_punto7)

learning_rate_init  accuracy
0               0.010  0.837989
1               0.001  0.821229

In [95]:
#----------------------------------------------------------
# Comparación de los accuracy del punto 6 y punto 7
#----------------------------------------------------------

# Se imprime la red neuronal con el mayor valor de Accuracy del punto 6
print("Accuracy de la red neuronal del punto 6:", max_accuracy["Accuracy"])

# Encontrar la red neuronal con el mayor valor de Accuracy
max_accuracy_p7 = max(resultados_p7, key=lambda x: x["accuracy"])

# Se imprime la red neuronal con el mayor valor de Accuracy del punto 7
print("Accuracy de la red neuronal del punto 7:", max_accuracy_p7["accuracy"])
print("    ")

if max_accuracy["Accuracy"] > max_accuracy_p7["accuracy"]:
    print("La red neuronal empeoró la exactitud, por lo que la mejor red tiene un accuracy de", round(max_accuracy["Accuracy"],4),
          "que corresponde al punto 6.")
    print("-------------------------------------------")
    print("Mejores hiperparámetros:")
    print("-------------------------------------------")
    print("Topología:", max_accuracy["Topología"])
    print("Función de activación:", max_accuracy["Función de activación"])
    print("Solver:", max_accuracy["Solver"])
    print("Accuracy:", max_accuracy["Accuracy"])
    print("Matriz de confusión:", max_accuracy["Matriz de confusión"])
    
elif max_accuracy["Accuracy"] < max_accuracy_p7["accuracy"]:
    print("La red neuronal mejoró la exactitud, por lo que la mejor red tiene un accuracy de", round(max_accuracy_p7["accuracy"],4), 
          "que corresponde al punto 7.")
    print("-------------------------------------------")
    print("Mejores hiperparámetros:")
    print("-------------------------------------------")
    print("learning_rate_init:", max_accuracy_p7["learning_rate_init"])
    print("Función de activación:", activation)
    print("Solver:", solver)
else:
    print("La red neuronal mantuvo la exactitud, ya que los valores de accuracy dieron igual aunque se haya realizado variaciones en los hiperparámetros.")



Accuracy de la red neuronal del punto 6: 0.8324
Accuracy de la red neuronal del punto 7: 0.8379888268156425
    
La red neuronal mejoró la exactitud, por lo que la mejor red tiene un accuracy de 0.838 que corresponde al punto 7.
-------------------------------------------
Mejores hiperparámetros:
-------------------------------------------
learning_rate_init: 0.01
Función de activación: relu
Solver: adam
